File: sdks/java/FastDataBrokerQuicClient.java
package com.fastdatabroker.sdk;

import java.io.*;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.Consumer;
import java.util.logging.Level;
import java.util.logging.Logger;

import com.google.gson.*;

/**
 * FastDataBroker Java SDK - QUIC with PSK Authentication
 * High-performance client library with Pre-Shared Key authentication
 * 
 * Version: 1.0.0
 * Protocol: QUIC 1.0 (RFC 9000)
 * Authentication: TLS 1.3 PSK (Pre-Shared Key)
 */

public class FastDataBrokerQuicClient {
    private static final Logger LOGGER = Logger.getLogger(FastDataBrokerQuicClient.class.getName());
    private static final Gson GSON = new Gson();

    public enum Priority {
        LOW(1), NORMAL(5), HIGH(10), CRITICAL(20);

        private final int value;
        Priority(int value) {
            this.value = value;
        }

        public int getValue() {
            return value;
        }
    }

    public enum ConnectionState {
        DISCONNECTED, CONNECTING, CONNECTED, AUTHENTICATED, ERROR
    }

    // Configuration
    public static class QuicConnectionConfig {
        public String host;
        public int port;
        public String tenantId;
        public String clientId;
        public String pskSecret;
        public String secrets;
        public int idleTimeoutMs = 30000;
        public int maxStreams = 100;
        public boolean autoReconnect = true;
        public int readTimeoutMs = 60000;

        public QuicConnectionConfig(String host, int port, String tenantId, String clientId, String pskSecret) {
            this.host = host;
            this.port = port;
            this.tenantId = tenantId;
            this.clientId = clientId;
            this.pskSecret = pskSecret;
        }
    }

    // Message envelope
    public static class Message {
        public String topic;
        public Object payload;
        public Priority priority = Priority.NORMAL;
        public int ttlSeconds = 3600;
        public Map<String, String> headers = new HashMap<>();

        public Message(String topic, Object payload) {
            this.topic = topic;
            this.payload = payload;
        }

        public Message withPriority(Priority priority) {
            this.priority = priority;
            return this;
        }

        public Message withTtl(int seconds) {
            this.ttlSeconds = seconds;
            return this;
        }

        public Message withHeader(String key, String value) {
            this.headers.put(key, value);
            return this;
        }
    }

    // Delivery result
    public static class DeliveryResult {
        public String messageId;
        public String status;
        public double latencyMs;
        public long timestamp;

        public DeliveryResult(String messageId, String status, double latencyMs, long timestamp) {
            this.messageId = messageId;
            this.status = status;
            this.latencyMs = latencyMs;
            this.timestamp = timestamp;
        }

        @Override
        public String toString() {
            return String.format("DeliveryResult{messageId='%s', status='%s', latencyMs=%.2f}", 
                messageId, status, latencyMs);
        }
    }

    // Connection statistics
    public static class ConnectionStats {
        public boolean isConnected;
        public long messagesSent;
        public long messagesReceived;
        public long connectionTimeMs;
        public long uptimeSeconds;
        public long lastMessageTime;

        public ConnectionStats(boolean isConnected, long messagesSent, long messagesReceived,
                             long connectionTimeMs, long uptimeSeconds, long lastMessageTime) {
            this.isConnected = isConnected;
            this.messagesSent = messagesSent;
            this.messagesReceived = messagesReceived;
            this.connectionTimeMs = connectionTimeMs;
            this.uptimeSeconds = uptimeSeconds;
            this.lastMessageTime = lastMessageTime;
        }

        @Override
        public String toString() {
            return String.format("ConnectionStats{isConnected=%b, messagesSent=%d, messagesReceived=%d, " +
                "connectionTimeMs=%d, uptimeSeconds=%d}", isConnected, messagesSent, messagesReceived,
                connectionTimeMs, uptimeSeconds);
        }
    }

    // ========================================================================
    // Client Implementation
    // ========================================================================

    private final QuicConnectionConfig config;
    private Socket socket;
    private PrintWriter writer;
    private BufferedReader reader;
    private boolean connected = false;
    private boolean authenticated = false;
    private ConnectionState state = ConnectionState.DISCONNECTED;
    private final Map<String, Consumer<JsonObject>> messageHandlers = new ConcurrentHashMap<>();
    private long connectionStart = 0;
    private ExecutorService executorService;
    private volatile boolean running = false;

    // Statistics
    private long messagesSent = 0;
    private long messagesReceived = 0;
    private long lastMessageTime = 0;
    private final Object statsLock = new Object();

    // Constructor
    public FastDataBrokerQuicClient(QuicConnectionConfig config) {
        this.config = config;
        LOGGER.info(String.format("Initialized FastDataBroker QUIC client for %s:%s",
            config.tenantId, config.clientId));
    }

    /**
     * Generate PSK identity and secret hash
     */
    private Map<String, String> generatePskIdentity() {
        String identity = config.tenantId + ":" + config.clientId;
        String secretHash = hashSha256(config.pskSecret);
        
        Map<String, String> result = new HashMap<>();
        result.put("identity", identity);
        result.put("secretHash", secretHash);
        return result;
    }

    /**
     * Compute SHA-256 hash
     */
    private String hashSha256(String input) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(input.getBytes(StandardCharsets.UTF_8));
            StringBuilder hexString = new StringBuilder();
            for (byte b : hash) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) hexString.append('0');
                hexString.append(hex);
            }
            return hexString.toString();
        } catch (Exception e) {
            throw new RuntimeException("Failed to hash secret", e);
        }
    }

    /**
     * Connect to FastDataBroker with PSK authentication
     */
    public synchronized void connect() throws IOException {
        if (connected) {
            LOGGER.warning("Already connected");
            return;
        }

        try {
            state = ConnectionState.CONNECTING;
            LOGGER.info(String.format("Connecting to %s:%d...", config.host, config.port));

            // Create socket
            socket = new Socket(config.host, config.port);
            socket.setKeepAlive(true);
            socket.setSoTimeout(config.readTimeoutMs);

            writer = new PrintWriter(new OutputStreamWriter(socket.getOutputStream(), StandardCharsets.UTF_8), true);
            reader = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8));

            connected = true;
            connectionStart = System.currentTimeMillis();
            LOGGER.info("✓ TCP connection established");

            // Send PSK handshake
            sendPskHandshake();

            // Start receive loop
            executorService = Executors.newSingleThreadExecutor(r -> {
                Thread t = new Thread(r, "FastDataBroker-Receive-Thread");
                t.setDaemon(true);
                return t;
            });

            running = true;
            executorService.execute(this::receiveLoop);

            authenticated = true;
            state = ConnectionState.AUTHENTICATED;
            LOGGER.info("✓ PSK authentication successful");

        } catch (IOException e) {
            state = ConnectionState.ERROR;
            connected = false;
            LOGGER.log(Level.SEVERE, "Connection failed", e);
            throw e;
        }
    }

    /**
     * Send PSK handshake
     */
    private void sendPskHandshake() {
        Map<String, String> psk = generatePskIdentity();
        JsonObject handshake = new JsonObject();
        handshake.addProperty("type", "psk_auth");
        handshake.addProperty("identity", psk.get("identity"));
        handshake.addProperty("secretHash", psk.get("secretHash"));
        handshake.addProperty("timestamp", System.currentTimeMillis());

        writer.println(handshake.toString());
        LOGGER.fine(String.format("Sent PSK handshake for %s", psk.get("identity")));
    }

    /**
     * Receive loop
     */
    private void receiveLoop() {
        String line;
        try {
            while (running && connected) {
                try {
                    line = reader.readLine();
                    if (line == null) {
                        LOGGER.info("Server closed connection");
                        connected = false;
                        break;
                    }

                    if (!line.isEmpty()) {
                        handleIncomingMessage(line);
                    }
                } catch (SocketTimeoutException e) {
                    continue;
                }
            }
        } catch (IOException e) {
            if (running && connected) {
                LOGGER.log(Level.WARNING, "Receive error", e);
            }
            connected = false;
        }
    }

    /**
     * Handle incoming message
     */
    private void handleIncomingMessage(String messageStr) {
        try {
            JsonObject parsed = JsonParser.parseString(messageStr).getAsJsonObject();

            if ("message".equals(parsed.get("type").getAsString())) {
                String topic = parsed.get("topic").getAsString();
                Consumer<JsonObject> handler = messageHandlers.get(topic);

                if (handler != null) {
                    handler.accept(parsed);
                    synchronized (statsLock) {
                        messagesReceived++;
                        lastMessageTime = System.currentTimeMillis();
                    }
                }
            }
        } catch (Exception e) {
            // Ignore malformed messages
        }
    }

    /**
     * Send message to FastDataBroker
     */
    public DeliveryResult sendMessage(Message message) throws IOException {
        if (!connected || socket == null) {
            throw new IOException("Not connected to FastDataBroker");
        }

        long startTime = System.currentTimeMillis();
        String messageId = String.format("msg_%d_%d", System.currentTimeMillis(), 
            (int)(Math.random() * 10000));

        JsonObject envelope = new JsonObject();
        envelope.addProperty("type", "message");
        envelope.addProperty("id", messageId);
        envelope.addProperty("topic", message.topic);
        envelope.add("payload", GSON.toJsonTree(message.payload));
        envelope.addProperty("priority", message.priority.getValue());
        envelope.addProperty("ttl", message.ttlSeconds);
        envelope.add("headers", GSON.toJsonTree(message.headers));
        envelope.addProperty("timestamp", System.currentTimeMillis());

        writer.println(envelope.toString());

        synchronized (statsLock) {
            messagesSent++;
            lastMessageTime = System.currentTimeMillis();
        }

        double latency = System.currentTimeMillis() - startTime;

        return new DeliveryResult(messageId, "success", latency, System.currentTimeMillis());
    }

    /**
     * Register message handler
     */
    public void onMessage(String topic, Consumer<JsonObject> handler) {
        messageHandlers.put(topic, handler);
        LOGGER.info("Registered handler for topic: " + topic);
    }

    /**
     * Unregister message handler
     */
    public void offMessage(String topic) {
        messageHandlers.remove(topic);
        LOGGER.info("Unregistered handler for topic: " + topic);
    }

    /**
     * Get connection statistics
     */
    public ConnectionStats getStats() {
        synchronized (statsLock) {
            long uptime = connected ? System.currentTimeMillis() - connectionStart : 0;
            return new ConnectionStats(
                connected,
                messagesSent,
                messagesReceived,
                uptime,
                uptime / 1000,
                lastMessageTime
            );
        }
    }

    /**
     * Check if connected
     */
    public boolean isConnected() {
        return connected && authenticated;
    }

    /**
     * Send multiple messages in parallel
     */
    public List<DeliveryResult> sendMessagesParallel(List<Message> messages, int numWorkers) {
        if (!isConnected()) {
            throw new RuntimeException("Not connected to FastDataBroker");
        }

        List<DeliveryResult> results = Collections.synchronizedList(new ArrayList<>());
        ExecutorService executor = Executors.newFixedThreadPool(numWorkers);
        List<Future<?>> futures = new ArrayList<>();

        for (Message message : messages) {
            Future<?> future = executor.submit(() -> {
                try {
                    DeliveryResult result = sendMessage(message);
                    results.add(result);
                } catch (Exception e) {
                    results.add(new DeliveryResult(
                        "error_" + System.currentTimeMillis(),
                        "failed",
                        0,
                        System.currentTimeMillis()
                    ));
                }
            });
            futures.add(future);
        }

        // Wait for all to complete
        for (Future<?> future : futures) {
            try {
                future.get();
            } catch (Exception e) {
                LOGGER.log(Level.WARNING, "Error waiting for message", e);
            }
        }

        executor.shutdown();
        return results;
    }

    /**
     * Send messages with progress callback
     */
    public List<DeliveryResult> sendMessagesParallelWithProgress(
        List<Message> messages,
        int numWorkers,
        Consumer<ProgressUpdate> callback
    ) {
        if (!isConnected()) {
            throw new RuntimeException("Not connected to FastDataBroker");
        }

        List<DeliveryResult> results = Collections.synchronizedList(new ArrayList<>());
        ExecutorService executor = Executors.newFixedThreadPool(numWorkers);
        AtomicInteger completed = new AtomicInteger(0);
        List<Future<?>> futures = new ArrayList<>();
        int total = messages.size();

        for (Message message : messages) {
            Future<?> future = executor.submit(() -> {
                try {
                    DeliveryResult result = sendMessage(message);
                    results.add(result);
                } catch (Exception e) {
                    results.add(new DeliveryResult(
                        "error_" + System.currentTimeMillis(),
                        "failed",
                        0,
                        System.currentTimeMillis()
                    ));
                }
                int count = completed.incrementAndGet();
                if (callback != null) {
                    callback.accept(new ProgressUpdate(count, total));
                }
            });
            futures.add(future);
        }

        // Wait for all to complete
        for (Future<?> future : futures) {
            try {
                future.get();
            } catch (Exception e) {
                LOGGER.log(Level.WARNING, "Error waiting for message", e);
            }
        }

        executor.shutdown();
        return results;
    }

    /**
     * Progress update for parallel processing
     */
    public static class ProgressUpdate {
        public final int completed;
        public final int total;

        public ProgressUpdate(int completed, int total) {
            this.completed = completed;
            this.total = total;
        }
    }

    /**
     * Disconnect from FastDataBroker
     */
    public synchronized void disconnect() {
        running = false;

        try {
            if (reader != null) reader.close();
            if (writer != null) writer.close();
            if (socket != null && !socket.isClosed()) socket.close();
        } catch (IOException e) {
            LOGGER.log(Level.WARNING, "Error during disconnect", e);
        }

        if (executorService != null) {
            executorService.shutdown();
            try {
                if (!executorService.awaitTermination(5, TimeUnit.SECONDS)) {
                    executorService.shutdownNow();
                }
            } catch (InterruptedException e) {
                executorService.shutdownNow();
                Thread.currentThread().interrupt();
            }
        }

        connected = false;
        authenticated = false;
        state = ConnectionState.DISCONNECTED;
        LOGGER.info("✓ Disconnected from FastDataBroker");
    }

    // ========================================================================
    // Static Factory Methods
    // ========================================================================

    public static FastDataBrokerQuicClient create(QuicConnectionConfig config) {
        return new FastDataBrokerQuicClient(config);
    }

    public static String getPskSecretFromEnv() {
        String secret = System.getenv("QUIC_PSK_SECRET");
        if (secret == null || secret.isEmpty()) {
            throw new IllegalStateException(
                "QUIC_PSK_SECRET environment variable not set. " +
                "Get it from: POST /api/quic/psks"
            );
        }
        return secret;
    }

    // ========================================================================
    // Example Usage
    // ========================================================================

    public static void main(String[] args) throws Exception {
        String pskSecret = System.getenv("QUIC_PSK_SECRET");
        if (pskSecret == null || pskSecret.isEmpty()) {
            pskSecret = "test-secret-key";
        }

        QuicConnectionConfig config = new QuicConnectionConfig(
            "localhost",
            6000,
            "test-tenant",
            "test-client",
            pskSecret
        );

        FastDataBrokerQuicClient client = new FastDataBrokerQuicClient(config);

        try {
            client.connect();
            System.out.println("✓ Client connected successfully");

            // Send message
            Message message = new Message("test.topic", new java.util.HashMap<>() {{
                put("data", "test");
            }}).withPriority(Priority.NORMAL);

            DeliveryResult result = client.sendMessage(message);
            System.out.println("✓ Message sent: " + result);

            // Get stats
            ConnectionStats stats = client.getStats();
            System.out.println("✓ Stats: " + stats);

            // Keep running
            Thread.sleep(10_000);

        } finally {
            client.disconnect();
        }
    }
}

// Import helper for SocketTimeoutException
import java.net.SocketTimeoutException;
import java.io.OutputStreamWriter;

File: sdks/csharp/FastDataBrokerQuicClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace FastDataBroker.SDK
{
    /// <summary>
    /// FastDataBroker C# SDK - QUIC with PSK Authentication
    /// High-performance client library with Pre-Shared Key authentication
    /// 
    /// Version: 1.0.0
    /// Protocol: QUIC 1.0 (RFC 9000)
    /// Authentication: TLS 1.3 PSK (Pre-Shared Key)
    /// </summary>
    public class FastDataBrokerQuicClient : IDisposable
    {
        // ====================================================================
        // Types and Enums
        // ====================================================================

        public enum Priority
        {
            Low = 1,
            Normal = 5,
            High = 10,
            Critical = 20
        }

        public enum ConnectionState
        {
            Disconnected,
            Connecting,
            Connected,
            Authenticated,
            Error
        }

        public class Message
        {
            public string Topic { get; set; }
            public object Payload { get; set; }
            public Priority Priority { get; set; } = Priority.Normal;
            public int TtlSeconds { get; set; } = 3600;
            public Dictionary<string, string> Headers { get; set; } = new();
        }

        public class DeliveryResult
        {
            public string MessageId { get; set; }
            public string Status { get; set; }
            public double LatencyMs { get; set; }
            public long Timestamp { get; set; }

            public override string ToString()
                => $"DeliveryResult{{MessageId='{MessageId}', Status='{Status}', LatencyMs={LatencyMs:F2}}}";
        }

        public class ConnectionStats
        {
            public bool IsConnected { get; set; }
            public long MessagesSent { get; set; }
            public long MessagesReceived { get; set; }
            public long ConnectionTimeMs { get; set; }
            public long UptimeSeconds { get; set; }
            public long LastMessageTime { get; set; }

            public override string ToString()
                => $"ConnectionStats{{IsConnected={IsConnected}, MessagesSent={MessagesSent}, " +
                   $"MessagesReceived={MessagesReceived}, ConnectionTimeMs={ConnectionTimeMs}}}";
        }

        public class QuicConnectionConfig
        {
            public string Host { get; set; }
            public int Port { get; set; }
            public string TenantId { get; set; }
            public string ClientId { get; set; }
            public string PskSecret { get; set; }
            public string Secrets { get; set; }
            public int IdleTimeoutMs { get; set; } = 30000;
            public int MaxStreams { get; set; } = 100;
            public bool AutoReconnect { get; set; } = true;
            public int ReadTimeoutMs { get; set; } = 60000;
        }

        // ====================================================================
        // Private Fields
        // ====================================================================

        private readonly QuicConnectionConfig _config;
        private TcpClient _client;
        private NetworkStream _stream;
        private StreamReader _reader;
        private StreamWriter _writer;
        private bool _connected = false;
        private bool _authenticated = false;
        private ConnectionState _state = ConnectionState.Disconnected;
        private readonly Dictionary<string, Action<JsonElement>> _messageHandlers = new();
        private long _connectionStart = 0;
        private CancellationTokenSource _cancellationTokenSource;
        private Task _receiveTask;
        private readonly object _statsLock = new();

        // Statistics
        private long _messagesSent = 0;
        private long _messagesReceived = 0;
        private long _lastMessageTime = 0;

        private static readonly JsonSerializerOptions JsonOptions = new()
        {
            PropertyNameCaseInsensitive = true,
            DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull
        };

        // ====================================================================
        // Constructor and Disposal
        // ====================================================================

        public FastDataBrokerQuicClient(QuicConnectionConfig config)
        {
            _config = config;
            Console.WriteLine($"Initialized FastDataBroker QUIC client for {config.TenantId}:{config.ClientId}");
        }

        public void Dispose()
        {
            Disconnect().Wait();
            _cancellationTokenSource?.Dispose();
            _reader?.Dispose();
            _writer?.Dispose();
            _stream?.Dispose();
            _client?.Dispose();
        }

        // ====================================================================
        // Connection Management
        // ====================================================================

        /// <summary>
        /// Generate PSK identity and secret hash
        /// </summary>
        private (string Identity, string SecretHash) GeneratePskIdentity()
        {
            var identity = $"{_config.TenantId}:{_config.ClientId}";
            var secretHash = ComputeSha256Hash(_config.PskSecret);
            return (identity, secretHash);
        }

        /// <summary>
        /// Compute SHA-256 hash
        /// </summary>
        private static string ComputeSha256Hash(string input)
        {
            using (var sha256 = SHA256.Create())
            {
                var hashedBytes = sha256.ComputeHash(Encoding.UTF8.GetBytes(input));
                return Convert.ToHexString(hashedBytes).ToLower();
            }
        }

        /// <summary>
        /// Connect to FastDataBroker with PSK authentication
        /// </summary>
        public async Task ConnectAsync()
        {
            if (_connected)
            {
                Console.WriteLine("Already connected");
                return;
            }

            try
            {
                _state = ConnectionState.Connecting;
                Console.WriteLine($"Connecting to {_config.Host}:{_config.Port}...");

                // Create TCP client
                _client = new TcpClient
                {
                    ReceiveTimeout = _config.ReadTimeoutMs,
                    SendTimeout = 5000
                };

                await _client.ConnectAsync(_config.Host, _config.Port);
                _stream = _client.GetStream();
                _reader = new StreamReader(_stream, Encoding.UTF8);
                _writer = new StreamWriter(_stream, Encoding.UTF8) { AutoFlush = true };

                _connected = true;
                _connectionStart = DateTime.UtcNow.Ticks / TimeSpan.TicksPerMillisecond;
                Console.WriteLine("✓ TCP connection established");

                // Send PSK handshake
                await SendPskHandshakeAsync();

                // Start receive loop
                _cancellationTokenSource = new CancellationTokenSource();
                _receiveTask = Task.Run(() => ReceiveLoop(_cancellationTokenSource.Token));

                _authenticated = true;
                _state = ConnectionState.Authenticated;
                Console.WriteLine("✓ PSK authentication successful");
            }
            catch (Exception ex)
            {
                _state = ConnectionState.Error;
                _connected = false;
                Console.WriteLine($"✗ Connection failed: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Send PSK handshake
        /// </summary>
        private async Task SendPskHandshakeAsync()
        {
            var psk = GeneratePskIdentity();
            var handshake = new
            {
                type = "psk_auth",
                identity = psk.Identity,
                secret_hash = psk.SecretHash,
                timestamp = DateTime.UtcNow.Ticks / TimeSpan.TicksPerMillisecond
            };

            var json = JsonSerializer.Serialize(handshake, JsonOptions);
            await _writer.WriteLineAsync(json);
        }

        /// <summary>
        /// Receive loop
        /// </summary>
        private async void ReceiveLoop(CancellationToken cancellationToken)
        {
            try
            {
                while (!cancellationToken.IsCancellationRequested && _connected)
                {
                    try
                    {
                        var line = await _reader.ReadLineAsync();
                        if (line == null)
                        {
                            Console.WriteLine("Server closed connection");
                            _connected = false;
                            break;
                        }

                        if (!string.IsNullOrEmpty(line))
                        {
                            HandleIncomingMessage(line);
                        }
                    }
                    catch (OperationCanceledException)
                    {
                        break;
                    }
                    catch (IOException) when (!_connected)
                    {
                        break;
                    }
                }
            }
            catch (Exception ex)
            {
                if (_connected)
                {
                    Console.WriteLine($"Receive error: {ex.Message}");
                }
                _connected = false;
            }
        }

        /// <summary>
        /// Handle incoming message
        /// </summary>
        private void HandleIncomingMessage(string messageStr)
        {
            try
            {
                var parsed = JsonDocument.Parse(messageStr).RootElement;

                if (parsed.TryGetProperty("type", out var typeEl) && typeEl.GetString() == "message")
                {
                    if (parsed.TryGetProperty("topic", out var topicEl))
                    {
                        var topic = topicEl.GetString();
                        if (_messageHandlers.TryGetValue(topic, out var handler))
                        {
                            handler(parsed);
                            lock (_statsLock)
                            {
                                _messagesReceived++;
                                _lastMessageTime = DateTime.UtcNow.Ticks / TimeSpan.TicksPerMillisecond;
                            }
                        }
                    }
                }
            }
            catch
            {
                // Ignore malformed messages
            }
        }

        // ====================================================================
        // Message Operations
        // ====================================================================

        /// <summary>
        /// Send message to FastDataBroker
        /// </summary>
        public async Task<DeliveryResult> SendMessageAsync(Message message)
        {
            if (!_connected || _client == null)
            {
                throw new InvalidOperationException("Not connected to FastDataBroker");
            }

            var startTime = DateTime.UtcNow.Ticks / TimeSpan.TicksPerMillisecond;
            var messageId = $"msg_{startTime}_{Random.Shared.Next(10000)}";

            var envelope = new
            {
                type = "message",
                id = messageId,
                topic = message.Topic,
                payload = message.Payload,
                priority = (int)message.Priority,
                ttl = message.TtlSeconds,
                headers = message.Headers,
                timestamp = DateTime.UtcNow.Ticks / TimeSpan.TicksPerMillisecond
            };

            var json = JsonSerializer.Serialize(envelope, JsonOptions);
            await _writer.WriteLineAsync(json);

            lock (_statsLock)
            {
                _messagesSent++;
                _lastMessageTime = DateTime.UtcNow.Ticks / TimeSpan.TicksPerMillisecond;
            }

            var latency = (DateTime.UtcNow.Ticks / TimeSpan.TicksPerMillisecond) - startTime;

            return new DeliveryResult
            {
                MessageId = messageId,
                Status = "success",
                LatencyMs = latency,
                Timestamp = DateTime.UtcNow.Ticks / TimeSpan.TicksPerMillisecond
            };
        }

        // ====================================================================
        // Message Handlers
        // ====================================================================

        /// <summary>
        /// Register message handler for topic
        /// </summary>
        public void OnMessage(string topic, Action<JsonElement> handler)
        {
            _messageHandlers[topic] = handler;
            Console.WriteLine($"Registered handler for topic: {topic}");
        }

        /// <summary>
        /// Unregister message handler
        /// </summary>
        public void OffMessage(string topic)
        {
            if (_messageHandlers.Remove(topic))
            {
                Console.WriteLine($"Unregistered handler for topic: {topic}");
            }
        }

        // ====================================================================
        // Connection Status
        // ====================================================================

        /// <summary>
        /// Get connection statistics
        /// </summary>
        public ConnectionStats GetStats()
        {
            lock (_statsLock)
            {
                var uptime = _connected 
                    ? (DateTime.UtcNow.Ticks / TimeSpan.TicksPerMillisecond) - _connectionStart 
                    : 0;

                return new ConnectionStats
                {
                    IsConnected = _connected,
                    MessagesSent = _messagesSent,
                    MessagesReceived = _messagesReceived,
                    ConnectionTimeMs = uptime,
                    UptimeSeconds = uptime / 1000,
                    LastMessageTime = _lastMessageTime
                };
            }
        }

        /// <summary>
        /// Check if connected
        /// </summary>
        public bool IsConnected => _connected && _authenticated;

        /// <summary>
        /// Send multiple messages in parallel
        /// </summary>
        public async Task<List<DeliveryResult>> SendMessagesParallelAsync(
            List<Message> messages,
            int numWorkers = 4
        )
        {
            if (!IsConnected)
            {
                throw new InvalidOperationException("Not connected to FastDataBroker");
            }

            var results = new List<DeliveryResult>();
            var options = new ParallelOptions { MaxDegreeOfParallelism = numWorkers };

            await Task.Run(() =>
            {
                Parallel.ForEach(messages, options, async message =>
                {
                    try
                    {
                        var result = await SendMessageAsync(message);
                        lock (results)
                        {
                            results.Add(result);
                        }
                    }
                    catch
                    {
                        lock (results)
                        {
                            results.Add(new DeliveryResult
                            {
                                MessageId = $"error_{DateTime.Now.Ticks}",
                                Status = "failed",
                                LatencyMs = 0,
                                Timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds()
                            });
                        }
                    }
                });
            });

            return results;
        }

        /// <summary>
        /// Send messages with progress tracking
        /// </summary>
        public async Task<List<DeliveryResult>> SendMessagesParallelWithProgressAsync(
            List<Message> messages,
            int numWorkers = 4,
            Action<int, int> onProgress = null
        )
        {
            if (!IsConnected)
            {
                throw new InvalidOperationException("Not connected to FastDataBroker");
            }

            var results = new List<DeliveryResult>();
            var completed = 0;
            var lockObj = new object();
            var options = new ParallelOptions { MaxDegreeOfParallelism = numWorkers };

            await Task.Run(() =>
            {
                Parallel.ForEach(messages, options, async message =>
                {
                    try
                    {
                        var result = await SendMessageAsync(message);
                        lock (lockObj)
                        {
                            results.Add(result);
                            completed++;
                            onProgress?.Invoke(completed, messages.Count);
                        }
                    }
                    catch
                    {
                        lock (lockObj)
                        {
                            results.Add(new DeliveryResult
                            {
                                MessageId = $"error_{DateTime.Now.Ticks}",
                                Status = "failed",
                                LatencyMs = 0,
                                Timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds()
                            });
                            completed++;
                            onProgress?.Invoke(completed, messages.Count);
                        }
                    }
                });
            });

            return results;
        }

        /// <summary>
        /// Synchronous version of SendMessagesParallelAsync
        /// </summary>
        public List<DeliveryResult> SendMessagesParallel(
            List<Message> messages,
            int numWorkers = 4
        )
        {
            return SendMessagesParallelAsync(messages, numWorkers).Result;
        }

        /// <summary>
        /// Synchronous version of SendMessagesParallelWithProgressAsync
        /// </summary>
        public List<DeliveryResult> SendMessagesParallelWithProgress(
            List<Message> messages,
            int numWorkers = 4,
            Action<int, int> onProgress = null
        )
        {
            return SendMessagesParallelWithProgressAsync(messages, numWorkers, onProgress).Result;
        }

        /// <summary>
        /// Disconnect from FastDataBroker
        /// </summary>
        public async Task DisconnectAsync()
        {
            _connected = false;
            _authenticated = false;
            _state = ConnectionState.Disconnected;

            _cancellationTokenSource?.Cancel();

            if (_receiveTask != null)
            {
                try
                {
                    await _receiveTask;
                }
                catch (OperationCanceledException)
                {
                    // Expected
                }
            }

            _stream?.Close();
            _client?.Close();
            Console.WriteLine("✓ Disconnected from FastDataBroker");
        }

        // Also provide synchronous version
        public void Disconnect()
        {
            DisconnectAsync().Wait();
        }

        // ====================================================================
        // Static Factory Methods
        // ====================================================================

        public static FastDataBrokerQuicClient Create(QuicConnectionConfig config)
        {
            return new FastDataBrokerQuicClient(config);
        }

        public static string GetPskSecretFromEnv()
        {
            var secret = Environment.GetEnvironmentVariable("QUIC_PSK_SECRET");
            if (string.IsNullOrEmpty(secret))
            {
                throw new InvalidOperationException(
                    "QUIC_PSK_SECRET environment variable not set. " +
                    "Get it from: POST /api/quic/psks"
                );
            }
            return secret;
        }

        // ====================================================================
        // Example Usage
        // ====================================================================

        public static async Task Main(string[] args)
        {
            var pskSecret = Environment.GetEnvironmentVariable("QUIC_PSK_SECRET") ?? "test-secret-key";

            var config = new QuicConnectionConfig
            {
                Host = "localhost",
                Port = 6000,
                TenantId = "test-tenant",
                ClientId = "test-client",
                PskSecret = pskSecret
            };

            using (var client = new FastDataBrokerQuicClient(config))
            {
                try
                {
                    await client.ConnectAsync();
                    Console.WriteLine("✓ Client connected successfully");

                    // Send message
                    var message = new Message
                    {
                        Topic = "test.topic",
                        Payload = new { data = "test" },
                        Priority = Priority.Normal
                    };

                    var result = await client.SendMessageAsync(message);
                    Console.WriteLine($"✓ Message sent: {result}");

                    // Get stats
                    var stats = client.GetStats();
                    Console.WriteLine($"✓ Stats: {stats}");

                    // Keep running
                    await Task.Delay(10000);
                }
                finally
                {
                    await client.DisconnectAsync();
                }
            }
        }
    }
}
